assembler/nasmlib.c

```diff
@@ -95,18 +94,11 @@ void *nasm_realloc (void *q, size_t size)
 	if (!q)
 		return nasm_malloc(size);
 
-	if (SEGMENT(q) == SEGMENT(&q) && size <= MAX_NEAR_ALLOC) /* near pointer and small re-alloc */
-	{
-		p = realloc(q, size);
-	}
-	else
-	{
-		p = nasm_malloc(size);
-		if (p)
-		{                    /* on fail, previous memory not freed */
-			memcpy(p, q, size);     /* FIXME copies too much!! */
-			nasm_free(q);
-		}
+	p = nasm_malloc(size);
+	if (p)
+	{                    /* on fail, previous memory not freed */
+		memcpy(p, q, size);     /* FIXME copies too much!! */
+		nasm_free(q);
 	}
```

Comment:
Do you think we might be better off in NASM to follow the idea you're using in C86? That is just redefine malloc/free/realloc to memalloc/memfree/memrealloc, which directs the function calls directly to mem.c?

In that way, nasmlib.c goes back to exactly how it was when you forked upstream, and all allocations are handled specially in mem.c, which would be ifdef ELKS for host/elks, and then easily replaced by libc code. Currently, when the mem.c code is placed in our C library, the NASM realloc isn't what it started with, but is handled by a slow malloc/free rather than actual realloc, for those systems that have it.